cachi2/core/package_managers/yarn_classic/workspaces.py

```diff
@@ -0,0 +1,114 @@
+import json
+from contextlib import suppress
+from itertools import chain
+from pathlib import Path
+from typing import Any, Iterable
+
+import pydantic
+
+from cachi2.core.errors import PackageRejected
+from cachi2.core.models.input import YarnClassicPackageInput
+from cachi2.core.rooted_path import RootedPath
+
+
+class Workspace(pydantic.BaseModel):
+    """Workspace model."""
+
+    path: Path  # path to a workspace.
+    package_contents: dict  # package data extracted from path/"package.json".
+    # package reference for potential nested workspace extraction:
+    package: YarnClassicPackageInput
+
+
+def ensure_no_path_leads_out(
+    paths: Iterable[Path],
+    source_dir: RootedPath,
+) -> None:
+    """Ensure no path leads out of source directory.
+
+    Raises an exception when any path is not relative to source directory.
+    Does nothing when path does not exist in the file system.
+    """
+    for path in paths:
+        if not path.is_relative_to(source_dir.path):
+            raise PackageRejected(
+                f"Found a workspace path which is not relative to package: {path}",
+                solution=(
+                    "Avoid using packages which try to access your filesystem "
+                    "outside of package directory."
+                ),
+            )
+
+
+def ensure_workspaces_are_well_formed(
+    paths: Iterable[Path],
+) -> None:
+    """Ensure that every workspace contains package.json.
+
+    Reject the package otherwise.
+    """
+    for p in paths:
+        if not Path(p, "package.json").is_file():
+            raise PackageRejected(
+                reason=f"Workspace {p} does not contain 'package.json'",
+                solution=None,
+            )
+
+
+def get_workspace_paths(
+    workspaces_globs: list[str],
+    source_dir: RootedPath,
+) -> Iterable[Path]:
+    """Resolve globs within source directory."""
+
+    def all_paths_matching(glob: str) -> list[Path]:
+        return [pth.resolve() for pth in source_dir.path.glob(glob)]
+
+    return chain.from_iterable(all_paths_matching(g) for g in workspaces_globs)
```

Comment:
stupid nitpick: Consider `map` instead of comprehension since you have the helper defined already and you have to chain the results? Might 0.01% more idiomatic :)

Reply:
> Sorry, I don't follow why map would be more idiomatic? I need to join several flat iterables into anther flat iterable, chain.from_iterable does exactly that. To the best of my knowledge map cannot do this.

Yes, I only meant the comprehension parth:
`chain.from_iterable(map(all_paths_matching), workspaces_globs)`

> there does not seem to be any good justification for using map instead of GC.

It's slightly more readable at times is all :) , why do you think I didn't even dare to call this a "nitpick" but a stupid one?

Reply:
Oh, I see it now! While I am not 100% agree with the "more readable" part I'll give it a try.